[WIP] Revamped octane tutorial (#1002)
ember-learn/guides-source: guides/release/tutorial/9.md

  <Jumbo>
    <h1>About Super Rentals</h1>
    <p>
      The Super Rentals website is a delightful project created to explore Ember.
      By building a property rental site, we can simultaneously imagine traveling AND building Ember applications.
    </p>
    <LinkTo @route="contact" class="button">Contact Us</LinkTo>
  </Jumbo>
</div>
```

```handlebars {data-filename="app/templates/contact.hbs" data-diff="-1,-2,+3,-19,+20"}
<div class="jumbo">
  <div class="right tomster"></div>
  <Jumbo>
    <h2>Contact Us</h2>
    <p>
      Super Rentals Representatives would love to help you<br>
      choose a destination or answer any questions you may have.
    </p>
    <address>
      Super Rentals HQ
      <p>
        1212 Test Address Avenue<br>
        Testington, OR 97233
      </p>
      <a href="tel:503.555.1212">+1 (503) 555-1212</a><br>
      <a href="mailto:superrentalsrep@emberjs.com">superrentalsrep@emberjs.com</a>
    </address>
    <LinkTo @route="about" class="button">About</LinkTo>
  </Jumbo>
</div>
```

After saving, everything should look exactly the same as before, and all the tests should still pass. Very nice!

<!-- TODO: screenshot of the page? -->

> While it may not save you a lot of characters in this case, it made the template slightly easier to read by *[encapsulating](TODO: link to encapsulating)* the implementation of the "jumbo" header into its own component, which allowed the reader to focus on things that are unique to just that page. Now, if we need to make a change to the header, we will just have to make it in a single place. Feel free to give that a try!

Reviewer: ```suggestion
While it may not save you a lot of characters in this case, [encapsulating](TODO: link to encapsulating)* the implementation of the "jumbo" header into its own component makes the template slightly easier to read, and it allows the reader to focus on things that are unique to just that page. Now, if we need to make a change to the header, we can make it in a single place. Feel free to give that a try!
```

import { module, test } from 'qunit';
import { setupRenderingTest } from 'ember-qunit';
import { render } from '@ember/test-helpers';
import hbs from 'htmlbars-inline-precompile';

module('Integration | Component | jumbo', function(hooks) {
  setupRenderingTest(hooks);

  test('it renders the content inside a jumbo header with a tomster', async function(assert) {
    await render(hbs`<Jumbo>Hello World</Jumbo>`);

    assert.dom('.jumbo').exists();
    assert.dom('.jumbo').containsText('Hello World');
    assert.dom('.jumbo .tomster').exists();
  });

  test('it renders', async function(assert) {
    // Set any properties with this.set('myProperty', 'value');
    // Handle any actions with this.set('myAction', function(val) { ... });

    await render(hbs`<Jumbo />`);

    assert.equal(this.element.textContent.trim(), '');

    // Template block usage:
    await render(hbs`
      <Jumbo>
        template block text
      </Jumbo>
    `);

    assert.equal(this.element.textContent.trim(), 'template block text');
  });
});
```

> Instead of navigating to a URL, we start the test by rendering our `<Jumbo>` component on the test page. This is useful because it may otherwise require a lot of setup and interaction just to get to a page where your component is used. Component tests allows us to skip all of that and focus exclusively on testing the component by itself.

Reviewer: ```suggestion
Instead of navigating to a URL, we start the test by rendering our `<Jumbo>` component on the test page. This is useful because it may otherwise require a lot of setup and interaction just to get to a page where your component is used. Component tests allows us to skip all of that and focus exclusively on testing the component in isolation.
```
